File: Prova di luglio 2019/soluzioni ale/es1-valoreMancante.c
// Alessandro Belfiore 975826
#include <stdio.h>
#include <stdlib.h>

int valoreMancante(int *A, int n) {
  if(A[0] != 0) {
    return 0;
  }
  if(A[n-1] != n) {
    return n;
  }
  int sx = 0, dx = n, m;
  while(sx < dx) {
    m = (sx + dx) / 2;
    if((A[m] != m) && (A[m-1] == m-1)) {
      return m;
    } else if(A[m] == m) {
      sx = m + 1;
    } else {
      dx = m;
    }
  }
  return 0;
}

int main( void ) {
  int n,num;
  scanf("%d",&n);
  int *A = malloc( n * sizeof(int) );
  for(int i = 0; i < n; i++) {
    scanf("%d",&num);
    A[i] = num;
  }
  printf("%d\n",valoreMancante(A,n));
  return 0;
}

File: Preparazione nuova (Estate 2023)/Esami/Esame del 3 luglio 2019/es1-valoreMancante.go
package main

import "fmt"

func valoreMancante(A []int, n int) int {
	dx := 0
	sx := n - 1
	for dx <= sx {
		m := (sx + dx) / 2
		if A[m]-A[dx] > m-dx {
			sx = m - 1
		} else {
			dx = m + 1
		}
	}
	return A[dx] - 1

}

func main() {
	A := []int{0, 1, 2, 3, 4, 6, 7}
	n := valoreMancante(A, len(A)+1)
	fmt.Println(n)
}
